### PR TITLE
Add Smooth L1 loss

### PR DESCRIPTION
**Pull request by softmaximalist**
### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

[Image quality metrics #4312](https://github.com/tracel-ai/burn/issues/4312)

### Changes

Implemented the Smooth L1 loss with a single struct field `beta`. Since Smooth L1 loss becomes L1 loss when beta = 0, beta must be set to a positive value. For beta = 0, users can use LpLoss (which has a constructor for L1 loss). I believe this is different from PyTorch which states in its documentation that beta must be non-negative. 

### Testing

Added 17 test cases covering different configurations and different forward methods.

### CODE REVIEW

File: crates/burn-nn/src/loss/smooth_l1.rs [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+
+        // Sort the dimensions to ascending order
+        let mut sorted_dims = dims.to_vec();
+        sorted_dims.sort();
+
+        // Reduce over specified dimensions
+        let mut result = error;
+        for &dim in sorted_dims.iter().rev() {
+            result = result.mean_dim(dim);
+        }
```
Comment (laggui) [inline crates/burn-nn/src/loss/smooth_l1.rs]: We actually have `tensor.mean_dims(..)` which could be used with the `sorted_dims`.

Comment (softmaximalist) [reply crates/burn-nn/src/loss/smooth_l1.rs]: I used this `mean_dim` approach in the implementation of `LpLoss` so I didn't notice it. I will update both smooth_l1.rs and lp_loss.rs to use `mean_dims` instead.

Comment (laggui) [reply crates/burn-nn/src/loss/smooth_l1.rs]: Didn't even realize in `LpLoss` :smile: thanks!
